### PR TITLE
Fix bug in solidity test artifact valdation

### PR DESCRIPTION
**Pull request by alcuadrado**
When running `test solidity --no-compile` without providing explicit test file we were running the validation of artifacts anyway.

This was unnecessary, and lead to some edge cases that we don't even need to handle. For example, test files that don't have any contract are roots, but don't have artifacts.

### CODE REVIEW

File: packages/hardhat/test/internal/builtin-plugins/solidity-test/task-action.ts [hunk 1]
```diff
@@ -484,6 +484,24 @@ describe("solidity-test/task-action", function () {
       }
     });
 
+    it("should not throw when noCompile is true and no testFiles are provided, even if the tests path covers uncompiled files", async () => {
```
Comment (marianfe) [inline packages/hardhat/test/internal/builtin-plugins/solidity-test/task-action.ts:487]: do we have the reverse of this: it("should throw when noCompile and testFiles are provided") ?

File: cspell.dictionary.txt [hunk 2]
```diff
@@ -123,6 +123,7 @@ txts
 ufixed
 Uints
 Uncontended
+uncompiled
```
Comment (marianfe) [inline cspell.dictionary.txt:126]: tiny nit: probably "not compiled" would express the same without adding a dictionary exception

**Review by marianfe — APPROVED** [review]

File: packages/hardhat/test/internal/builtin-plugins/solidity-test/task-action.ts (hunk 1 above)
Comment (alcuadrado) [reply packages/hardhat/test/internal/builtin-plugins/solidity-test/task-action.ts:487]: yes, that one was already present

File: cspell.dictionary.txt (hunk 2 above)
Comment (alcuadrado) [reply cspell.dictionary.txt:126]: good idea

File: packages/hardhat/src/internal/builtin-plugins/solidity-test/task-action.ts [hunk 3]
```diff
@@ -141,8 +141,11 @@ const runSolidityTests: NewTaskActionFunction<TestActionArguments> = async (
     ({ edrArtifactsWithMetadata, allBuildInfosAndOutputs } =
       await loadArtifacts(hre.solidity, ["contracts"]));
 
-    // When noCompile, validate selected test roots have compiled artifacts
-    if (noCompile === true) {
+    // When noCompile is enabled, only validate compiled artifacts for test roots
+    // explicitly selected by the user. If no files were specified, skip
+    // validating every discovered test root and run whatever compiled test
+    // suites are available.
+    if (noCompile === true && testFiles.length > 0) {
```
Comment (Wodann) [inline packages/hardhat/src/internal/builtin-plugins/solidity-test/task-action.ts:148]: Question: Is there a reason `hardhat compile` doesn't generate all possible test compilation artifacts?

If `hardhat test --no-compile` is called after `hardhat compile`, I would expect the check for compiled artifacts to pass with or without filter.
